### DISCUSSION TITLE
How to define local network router's IP in forwarding-rules not hardcoding it to particular value?

### DISCUSSION THREAD

**Discussion by kkovaletp** [DNSCrypt/dnscrypt-proxy#2368] [Q&A]
Hi Community and Team,

I'm new to the dnscrypt-proxy and set it up for 1st time y-day on my Mac to use all advantages of DoH across the system and not only in browsers.
However, I have other devices in my local network, so I had to configure forwarding-rules to let dnscrypt-proxy forward all requests for that devices to my router directly, as advised in the documentation. It works fine at home, but if I get my Mac to my friend and join his WiFi, it stops working, as he has configured his router with different IP.

My question is: is there a way to not hardcode the router's IP in that config and use some sort of dynamic router discovery or IP resolution? The OS knows it anyway, so I assume there might be some option to put a variable into the config, which then can be resolved to the router's IP dynamically at the time, each forwarding happens...

If this is a known question, please point me to the answer, as I didn't find it.
If for some reason it is not possible, as not been implemented yet, please let me know and I'll submit a feature request for that.

Waiting for your thoughts.
Thanks in advance,
Kostiantyn.

**Comment by lifenjoiner** [comment]
This is an interesting feature that I also imagined before. But AFAIK, there is no this feature implemented, nor in other programs I've ever seen.

The OS knows a dynamic DNS from the router according Dynamic Host Configuration Protocol (DHCP).
Once, you use a specific DNS, you had set your device's DNS static. I don't know if the OS still has the uptodate DNS from the DHCP server.
If not, extra DHCP library will be required. Can a single program claim as a DHCP client and send DHCP messages? Safe?

**Reply by ZzZombo** [reply]
> Can a single program claim as a DHCP client and send DHCP messages? Safe?

It should be. In fact, that's how WPAD using DHCP is supposed to be implemented, there is no way other than by sending a specific DHCP query and listening to the response.

**Reply by kkovaletp** [reply]
> The OS knows a dynamic DNS from the router according Dynamic Host Configuration Protocol (DHCP).

DHCP is (AFAIK) the most popular LAN configuration way and local DNS discovery using DHCP looks like a straightforward approach in this case. However, the OS already has that info and in theory, the app can send the request to the OS to get it.

What if there is no DHCP and we have manual network configuration for some reason? I don't know whether this case is even could be found in real life these days... But technically it is possible. What options do we have here? If we have manual network configuration, asking the OS looks like the best option, but there is no guarantee that the gateway has DNS on the same IP (as well as we couldn't be sure that the DNS even exists in this LAN). What will happen if the app will send some broadcast request in the LAN and if there will be a response, that IP would be considered a DNS? What could we do if there is no DNS at all and network nodes use some simple network lookup / WINS for name resolution? Could the app have some small and simple DNS to handle that?

These are just my thoughts... Maybe they will help you to better see the area and find a solution...

**Comment by lifenjoiner** [comment]
Let's do some experiments first:
[Try this CLI demo](https://github.com/lifenjoiner/dhcpdns/releases) to see if it works for you. 
If there is no binaries for your platform, compile it yourself with the `dev` branch.

**Reply by kkovaletp** [reply]
> Let's do some experiments first: [Try this CLI demo](https://github.com/lifenjoiner/dhcpdns/releases) to see if it works for you. If there is no binaries for your platform, compile it yourself with the `dev` branch.

1st of all, what should I expect to get as a result after running this project? how would you consider if it was successful or not?

then, can I run it from VirtualBox VM? Linux (which distro?) or Windows 11? Because I'm not an expert in GO and I don't feel ok with running some unknown experimental code on my main system.

**Reply by lifenjoiner** [reply]
Descriptions have been updated.
See if it gets the DNS you expect.

**Reply by kkovaletp** [reply]
> Descriptions have been updated. See if it gets the DNS you expect.

I've tested this script on my Windows 11 and Ubuntu 22.10 VirtualBox VMs and the output looks similar to the example on the release page, however because of the VirtualBox network topology, the script reported VirtualBox's virtual GW IP instead of my network's real GW. 

So, I tried to build the project on my Mac and run it on the physical system. To do that I installed Go by the brew (version 1.20.3 darwin/amd64) and after I executed the `go build .` command (because the `go build cli` returns an error that the `package cli is not in GOROOT`) from the project's root folder, I got the next error:
`./sockopt_others.go:40:15: undefined: syscall.BindToDevice`

My googling resulted in advice to replace the `syscall.BindToDevice` with some other function (e.g. `syscall.SetsockoptString`), but I have no experience in Go programming, so I don't know how the project needs to be modified to fix the issue and do exactly what was designed.

@lifenjoiner, could you please fix the issue and let me know?

**Reply by lifenjoiner** [reply]
Thanks for testing. Fixed, and CI shows it will work. @kkovaletp Binaries for Mac attached.

> reported VirtualBox's virtual GW IP instead of my network's real GW

It should keep the same page as the OS can get automatically, rather than get out of the box.

**Reply by kkovaletp** [reply]
> Fixed, and CI shows it will work

Thanks, I tried the darwin binaries and there is no developer signature, so macOS didn't allow it to run.
So I had to run it from source code and here is what I get:

```
2023/04/27 15:51:37 Targeting: [2001:4860:4860::8888]:80
2023/04/27 15:51:40 error: read udp6 [fe80::1060:fcf2:9ab4:8463%en8]:546: i/o timeout
2023/04/27 15:51:40 Targeting: 8.8.8.8:80
2023/04/27 15:51:40 error: listen udp4 :68: bind: address already in use
```

**Reply by lifenjoiner** [reply]
On Mac, [here](https://github.com/lifenjoiner/dhcpdns/actions/runs/4831147271/jobs/8608200905#step:5:20) shows 7/10 got.
There seems to be a system DHCP client occupies the `:68` exclusively, at OS up or when the network state changed, maybe aperiodic.

**Reply by kkovaletp** [reply]
> occupies the `:68` exclusively

I don't know for sure, but I have DNScrypt-proxy installed on my Mac and all DNS requests are forwarded through it. Maybe that is the reason.
BTW, after that test, I had to reboot my Mac, because after some time I noticed that new DNS requests were failing and DNScrypt-proxy's query log contains a lot of requests with NETWORK ERROR.
After reboot it is fine. Maybe your experimental code did some config changes, which were not reverted on process exit...

**Reply by lifenjoiner** [reply]
The demo does nothing special more than the package name saying: DHCP communication, requesting the DHCP server (router) for the DNS.

> BTW, after that test, I had to reboot my Mac, because after some time I noticed that new DNS requests were failing and DNScrypt-proxy's query log contains a lot of requests with NETWORK ERROR.

If not all requests failed or if you can still visit the web, that should not be an issue caused by this.
If you lost your connection completely, that will be the problem, and just reconnect will solve it I guess.

I don't have a Mac to test.
On Windows and Android, the connection is unaffected.

**Reply by kkovaletp** [reply]
> If not all requests failed or if you can still visit the web, that should not be an issue caused by this.
> If you lost your connection completely, that will be the problem, and just reconnect will solve it I guess.

There is a DNS cache, so all typical hosts are already known and no requests were sent for several hours, that is why I didn't notice the problem right after the testing.
I have wired and wireless connections and tried to reconnect both of them (subsequently and at the same time) with no effect, also I tried to restart DNScrypt-proxy, but only the system reboot helped.

**Reply by lifenjoiner** [reply]
A potential issue should be fixed. That's all about DHCP.
On Mac, it seems not as reliable as on Ubuntu nor Windows.

**Comment by lifenjoiner** [comment]
One more step, here is [an alpha version extending forwarding on my branch](https://github.com/lifenjoiner/dnscrypt-proxy/releases).
Anyone is interested can have a try and help improve it.

**Reply by dodmi** [reply]
Hello, thank you for implementing this. This sounds really great and helpfull. I'll gladly test it, as I have different problems with some public WLANs, using captive portals, where the portal site is only resolvable from the provided DNS.

**Reply by dodmi** [reply]
I just want to add: It seems to work great on my local network; platform is Windows x64

**Reply by dodmi** [reply]
After some testing: It works as intended for all other networks. Thank you ;)
Only problem with captive portals: Any domain lookup will be redirected to the captive portal, so you need to open the captive portal manually, since we don't want to send any lookup to the DHCP dns server.

**Comment by lifenjoiner** [comment]
More tests appreciated:
1. IPv6 DHCP. I have a buggy router from an ISP, which provides IPv6 occasionally and limited DHCP.
2. [Running as a non-root user](https://github.com/DNSCrypt/dnscrypt-proxy/wiki/Installation-linux#running-as-a-non-root-user) on Linux.

[An alpha version extending forwarding on my branch](https://github.com/lifenjoiner/dnscrypt-proxy/releases).

**Comment by designgrill** [comment]
Thanks for working on this @lifenjoiner. Any idea if this is getting merged any time soon? 

This would be a very useful feature, especially for Captive Portals whose IPs are not known before hand. I feel it is much easier to put a domain through the DHCP DNS than to try to figure out the IP behind a captive portal without compromising the rest of the traffic.

**Reply by lifenjoiner** [reply]
I respect Frank as decision maker, and also keep radical features in my branch.
